File: lib/random-words.rb
def reverse(word)
  letters = word.split('')

  letters
    .each_with_index
    .map { |_, idx| letters[-1 - idx] }
    .join
end

def palindrome?(word)
  word == reverse(word)
end

def random_word
  letters = 'abcdefghijklmnopqrstuvxz'.split('')
  size    = [3, 4, 5].sample

  (1..size).reduce('') do |acc, _|
    random_idx = (0..23).to_a.sample

    acc << letters[random_idx]
    acc
  end
end

def palindromes_from(limit)
  limit
    .times
    .map { random_word }
    .select { |word| palindrome?(word) }
end

puts reverse('leandro')
puts palindrome?('leandro')
puts palindrome?('ana')
puts random_word
puts palindromes_from(1000)

File: lib/random-words.js
const reverse = (word) => {
  const letters = word.split('')
  const lastIdx = letters.length - 1

  return letters
    .map((letter, idx) => { return letters[lastIdx - idx] })
    .join('')
}

const isPalindrome = (word) => {
  return word === reverse(word)
}

const randomMax = (sizeLimit) => {
  return Math.floor(
    Math.random() * sizeLimit
  )
}

const sample = (list) => {
  const randomIdx = randomMax(list.length)

  return list[randomIdx]
}

const iterateTimes = (times) => {
  return Array.from(Array(times).keys())
}

const randomWord = () => {
  const alphabet = 'abcdefghijklmnopqrstuvxz'.split('')
  const size     = sample([3, 4, 5])

  return iterateTimes(size)
    .reduce((acc) => { return acc.concat(sample(alphabet)) }, '')
}

const palindromesFrom = (limit) => {
  return iterateTimes(limit)
    .map(()        => { return randomWord() })
    .filter((word) => { return isPalindrome(word) })
}

console.log(reverse('leandro'))
console.log(isPalindrome('leandro'))
console.log(isPalindrome('ana'))
console.log(sample([3, 4, 5]))
console.log(randomWord())
console.log(palindromesFrom(1000))
